source/opt/mem_pass.cpp

```diff
@@ -92,21 +97,25 @@ bool MemPass::IsPtr(uint32_t ptrId) {
 ir::Instruction* MemPass::GetPtr(uint32_t ptrId, uint32_t* varId) {
   *varId = ptrId;
   ir::Instruction* ptrInst = get_def_use_mgr()->GetDef(*varId);
-  while (ptrInst->opcode() == SpvOpCopyObject) {
-    *varId = ptrInst->GetSingleWordInOperand(kCopyObjectOperandInIdx);
-    ptrInst = get_def_use_mgr()->GetDef(*varId);
+  ir::Instruction* varInst;
+
+  if (ptrInst->opcode() != SpvOpVariable &&
+      ptrInst->opcode() != SpvOpFunctionParameter) {
+    varInst = ptrInst->GetBaseAddress();
+  } else {
+    varInst = ptrInst;
   }
-  ir::Instruction* varInst = ptrInst;
-  while (varInst->opcode() != SpvOpVariable &&
-         varInst->opcode() != SpvOpFunctionParameter) {
-    if (IsNonPtrAccessChain(varInst->opcode())) {
-      *varId = varInst->GetSingleWordInOperand(kAccessChainPtrIdInIdx);
-    } else {
-      assert(varInst->opcode() == SpvOpCopyObject);
-      *varId = varInst->GetSingleWordInOperand(kCopyObjectOperandInIdx);
-    }
-    varInst = get_def_use_mgr()->GetDef(*varId);
+  if (varInst->opcode() == SpvOpVariable) {
+    *varId = varInst->result_id();
+  } else {
+    *varId = 0;
+  }
+
+  while (ptrInst->opcode() == SpvOpCopyObject) {
```

Comment:
At first I was concerned that functionality might have been changed by moving this down. But the GetBasePointer logic traces through OpCopyObject.   And I think the rest checks out.

So this is fine by me.